shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/util/MigrationDistributedCountDownLatch.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.shardingsphere.mode.repository.cluster.listener.DataChangedEvent;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+@Slf4j
+public final class MigrationDistributedCountDownLatch {
+    
+    private static final MigrationDistributedCountDownLatch INSTANCE = new MigrationDistributedCountDownLatch();
+    
+    private static final Pattern BARRIER_MATCH_PATTERN = Pattern.compile(DataPipelineConstants.DATA_PIPELINE_ROOT + "/(j\\d{2}[0-9a-f]+)/barrier/(enable|distable)/\\d+");
+    
+    private final ClusterPersistRepository clusterPersistRepository;
+    
+    private final Map<String, InnerCountDownLatchHolder> countDownLatchMap = new ConcurrentHashMap<>();
+    
+    private MigrationDistributedCountDownLatch() {
+        clusterPersistRepository = (ClusterPersistRepository) PipelineContext.getContextManager().getMetaDataContexts().getPersistService().getRepository();
+    }
+    
+    /**
+     * Get instance.
+     *
+     * @return instance
+     */
+    public static MigrationDistributedCountDownLatch getInstance() {
+        return INSTANCE;
+    }
+    
+    /**
+     * Register count down latch.
+     *
+     * @param parentPath parent path
+     * @param totalCount total count
+     */
+    public void register(final String parentPath, final int totalCount) {
+        clusterPersistRepository.persist(parentPath, "");
+        countDownLatchMap.computeIfAbsent(parentPath, k -> new InnerCountDownLatchHolder(totalCount, new CountDownLatch(1)));
+    }
+    
+    /**
+     * Persist ephemeral children node.
+     *
+     * @param parentPath parent path
+     * @param shardingItem sharding item
+     */
+    public void persistEphemeralChildrenNode(final String parentPath, final int shardingItem) {
+        String key = String.join("/", parentPath, Integer.toString(shardingItem));
+        clusterPersistRepository.delete(key);
+        clusterPersistRepository.persistEphemeral(key, "");
+    }
+    
+    /**
+     * Persist ephemeral children node.
+     *
+     * @param parentPath parent path
+     */
+    public void removeParentNode(final String parentPath) {
+        clusterPersistRepository.delete(String.join("/", parentPath));
+        countDownLatchMap.remove(parentPath);
+    }
+    
+    /**
+     * Await unitl all children node is ready.
+     *
+     * @param parentPath parent path
+     * @param timeout timeout
+     * @param timeUnit time unit
+     * @return
+     * true if the count reached zero and false if the waiting time elapsed before the count reached zero
+     */
+    public boolean await(final String parentPath, final long timeout, final TimeUnit timeUnit) {
+        InnerCountDownLatchHolder holder = countDownLatchMap.get(parentPath);
+        if (holder == null) {
+            return false;
+        }
```

Comment:
1, `holder == null` could be `null ==handler`

2, It's better to log before return false